src/Backend/Lua/Builtin.hs

```diff
@@ -118,30 +118,62 @@ builtins =
           return var[1]
         end
       |] )
-  , ( vStrVal, "__builtin_strval", [], Nothing
+  , ( vStrVal, "__builtin_strval", [], Just (1, \[l] -> (mempty, [ [lua|%l|] ]))
     , [luaStmts|
         local function __builtin_strval(x)
           return x
         end
       |] )
-  , ( vKSTR, "__builtin_kstr", [], Nothing
+  , ( vKSTR, "__builtin_kstr", [], Just (1, \[l] -> (mempty, [ [lua|%l|] ]))
     , [luaStmts|
         local function __builtin_kstr(x)
           return x
         end
       |] )
-  , ( vIntVal, "__builtin_intval", [], Nothing
+  , ( vIntVal, "__builtin_intval", [], Just (1, \[l] -> (mempty, [ [lua|%l|] ]))
     , [luaStmts|
         local function __builtin_intval(x)
           return x
         end
       |] )
-  , ( vKINT, "__builtin_kint", [], Nothing
+  , ( vKINT, "__builtin_kint", [], Just (1, \[l] -> (mempty, [ [lua|%l|] ]))
     , [luaStmts|
         local function __builtin_kint(x)
           return x
         end
       |] )
+  , ( vROWCONS, "__builtin_rowcons", [], Just (1, \[l] -> (mempty, [ [lua|%l|] ]))
+    , [luaStmts|
+        local function __builtin_rowcons(x)
+          return x
+        end
+      |] )
+  , ( vExtend, "__builtin_extend", [], Nothing
+    , [luaStmts|
+        local function __builtin_extend(key, value, record)
+          local new = {}
+          for k, v in pairs(record) do
+            new[k] = v
+          end
+          new[key] = value
+          return new
+        end
+      |] )
+  , ( vRestrict, "__builtin_restrict", [], Just (2, \[key, t] -> (mempty, [[lua| { _1 = %t[%key], _2 = %t }|]]))
```

Comment:
Ok, this inlined version is super dubious, but I agree with your reasoning. Though I dunno if it'll cause issues with Lua interop in the future.

Reply:
That's true, the FFI could observe our terrible optimisation, but I claim that's Undefined Behaviour.